File: 0338-counting-bits/0338-counting-bits.cpp
class Solution {
public:
    vector<int> countBits(int n) {
        vector<int> ans;
        for(int i = 0; i <= n; i++){
            bitset<32> tmp(i);
            ans.push_back(tmp.count());
        }
        return ans;

        //  =============================
        //  https://leetcode.com/problems/counting-bits/solutions/1808016/c-vectors-only-easy-to-understand-full-explanation

        //  O(nlogn)
        // vector<int> ans;
        // for(int i = 0; i <= n; i++){
        //     int sum = 0;
        //     int num = i;
        //     //  2진수이기 때문에 2로 나누면서 1이 남은경우 더해서 1의 개수를 파악
        //     while(num != 0){
        //         sum += num%2;
        //         num /= 2;
        //     }
        //     ans.push_back(sum);
        // }
        // return ans;

        //  O(n)
        // vector<int> t(n+1);
        // t[0] = 0;
        // for(int i = 1; i <= n; i++){
        //     //  2진수 이기 때문에 i는 i/2의 계산 값에 i%2( 0 또는 1) 을 더한다
        //     t[i] = t[i/2] + i%2;
        // }
            
        // return t;
    }
};

File: 0338-counting-bits/0338-counting-bits.cs
public class Solution {
    public int[] CountBits(int n) {
        //  https://leetcode.com/problems/counting-bits/solutions/3986178/97-97-dp-bit-manipulation-offset
        //  방법 자체는 c++에서 본 솔루션 방법
        //  주석되있는식은 같은 결과지만 비트 연산을 이용한 다른 방법
        int[] ans = new int[n+1];
        ans[0] = 0;
        for(int i = 1; i <= n; i++){
            ans[i] = ans[i/2] + i%2;
            //  ans[i] = ans[i >> 2] + (i & 1)
        }
        return ans;
    }
}